torchtune/modules/transformer.py

```diff
@@ -240,9 +254,11 @@ def forward(
             # in most cases input_pos_len should be 1
             mask = self.causal_mask[None, input_pos]
 
-        for layer in self.layers:
+        for i, layer in enumerate(self.layers):
             # shape: [b, s, d]
-            h = layer(h, mask=mask, input_pos=input_pos)
+            h = self.layer_dropouts[i](layer, h, mask=mask, input_pos=input_pos)
```

Comment:
I'm wondering if there's a world where we can actually do all of this without even touching `TransformerDecoder`'s forward (except for the change to support `output_hidden_states`). Suppose we generalize the init arg `layer` in `TransformerDecoder` to take in a list of `nn.Module`s. Then we provide functions to build a list of layers and call those whenever we're initializing this class instead. So in the existing case it'd look like

```
def get_decoder_layers(
	attn: CausalSelfAttention,
	mlp: nn.Module,
	sa_norm: nn.Module,
	mlp_norm: nn.Module,
	n_layers: int
) -> List[TransformerDecoderLayer]:
	return [TransformerDecoderLayer(attn, mlp, sa_norm, mlp_norm) for _ in n_layers]
```

And for the `LayerDropout` version:

```
def get_layer_skip_decoder_layers(
	attn: CausalSelfAttention,
	mlp: nn.Module,
	sa_norm: nn.Module,
	mlp_norm: nn.Module,
	n_layers: int,
        layer_dropout_prob: float = 0.5,
        layer_dropout_prob_layer_scale: str = "exp",
        layer_dropout_str: str = ":",
) -> List[LayerDropout]:
	layers = [TransformerDecoderLayer(attn, mlp, sa_norm, mlp_norm) for _ in n_layers]
	return create_layer_dropout_modules(
		layers, 
		layer_dropout_prob, 
		layer_dropout_prob_layer_scale,
		layer_dropout_str
	)
```

I am making one big assumption here: that we can move the callable `function` from the `LayerDropout` `forward` into its `__init__`. Then a given `LayerDropout` module would correspond to a single `nn.Module` it is wrapping. I guess this is less flexible from the `LayerDropout` perspective, but I think it should help from a composability perspective. (I am also implicitly changing the signature of `create_layer_dropout_modules` to account for this change, replacing `n_layers` -> `layers`.) Let me know if this assumption makes sense (it's possible there are important details of the implementation I'm overlooking with this proposal).

If this does work, we would expose things in our model builders by replacing e.g. [this](https://github.com/pytorch/torchtune/blob/74fb5e4cceb6e64473608106b77f9f9308ec11c3/torchtune/models/llama3/_component_builders.py#L100-L105) with calls to the appropriate builder. This is similar to what we do for some of our LoRA components (see e.g. [here](https://github.com/pytorch/torchtune/blob/74fb5e4cceb6e64473608106b77f9f9308ec11c3/torchtune/models/llama3/_component_builders.py#L211-L221)).

Also cc @pbontrager since this is related to ideas we've been talking about for cross-attention support.

Reply:
I like this idea, and I will try to implement it.
One concern I have is if `LayerDropout` wraps `TransformerDecoderLayer`, then this means the weights of the model will look something like `layer_dropout0.layer.attn.wk`, right? Then this will cause problems if we want to load to or from a model that doesn't have dropout as its weights' names will be something like `layer0.atten.wk`.

Reply:
LayerDropout has been re-implemented in the PR by creating a wrapper module. Please take a look.